### PR TITLE
Implement `kubectl alpha diff` to diff resources

### PR DESCRIPTION
**Pull request by apelisse**
`kubectl alpha diff` lets you diff your resources against live
resources, or last applied, or even preview what changes are going to be
applied on the cluster.

This is still quite premature, and mostly untested.

**What this PR does / why we need it**: 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:
Clearly not ready for Release note.
```release-note
NONE
```

kubernetes/community#287

### CODE REVIEW

**Comment by cblecker** [comment]
```
I1013 22:32:26.065] Verifying hack/make-rules/../../hack/verify-govet.sh
W1013 22:36:13.297] pkg/kubectl/apply/parse/factory.go:88: k8s.io/kubernetes/pkg/kubectl/apply.ListElementData composite literal uses unkeyed fields
W1013 22:36:13.298] pkg/kubectl/apply/parse/factory.go:91: k8s.io/kubernetes/pkg/kubectl/apply.MapElementData composite literal uses unkeyed fields
W1013 22:36:13.298] pkg/kubectl/apply/parse/factory.go:98: k8s.io/kubernetes/pkg/kubectl/apply.MapElementData composite literal uses unkeyed fields
W1013 22:36:13.298] pkg/kubectl/apply/parse/list_element.go:156: k8s.io/kubernetes/pkg/kubectl/apply.ListElement composite literal uses unkeyed fields
W1013 22:36:13.298] pkg/kubectl/apply/parse/map_element.go:42: k8s.io/kubernetes/pkg/kubectl/apply.MapElement composite literal uses unkeyed fields
W1013 22:36:13.299] pkg/kubectl/apply/parse/primitive_element.go:24: k8s.io/kubernetes/pkg/kubectl/apply.PrimitiveElement composite literal uses unkeyed fields
W1013 22:36:13.299] pkg/kubectl/apply/parse/type_element.go:41: k8s.io/kubernetes/pkg/kubectl/apply.TypeElement composite literal uses unkeyed fields
W1013 22:36:13.302] exit status 1
W1013 22:36:13.392] pkg/kubectl/apply/strategy/utils_test.go:47: k8s.io/kubernetes/pkg/kubectl/apply/parse.Factory composite literal uses unkeyed fields
W1013 22:36:13.393] exit status 1
W1013 22:36:14.245] pkg/kubectl/cmd/diff.go:224: k8s.io/kubernetes/pkg/kubectl/apply/parse.Factory composite literal uses unkeyed fields
W1013 22:36:14.316] exit status 1
W1013 22:37:14.059] make[1]: *** [vet] Error 1
I1013 22:37:14.160] Makefile:364: recipe for target 'vet' failed
I1013 22:37:14.160] FAILED   hack/make-rules/../../hack/verify-govet.sh	288s
```
```
I1013 22:28:29.527] List of generated docs doesn't match a freshly built list. Please run hack/update-generated-docs.sh
I1013 22:28:29.533] FAILED   hack/make-rules/../../hack/verify-generated-docs.sh	121s
```

**Comment by pwittrock** [comment]
Review status: 0 of 3 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/kubectl/cmd/diff.go, line 44 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KwaGHpMrV0R7Bx7K0FB:-KwaGHpMrV0R7Bx7K0FC:b-hho97p) ([raw file](https://github.com/kubernetes/kubernetes/blob/42a1831ea571eca3af6ea889bbf02f9553a3160f/pkg/kubectl/cmd/diff.go#L44)):*
> ```Go
> 
> var (
> 	diffLong    = templates.LongDesc(i18n.T(`TODO`))
> ```

Finish these

---

*[pkg/kubectl/cmd/diff.go, line 56 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KwaGOh0HkEs5El7oY7H:-KwaGOh0HkEs5El7oY7I:bczn572) ([raw file](https://github.com/kubernetes/kubernetes/blob/42a1831ea571eca3af6ea889bbf02f9553a3160f/pkg/kubectl/cmd/diff.go#L56)):*
> ```Go
> 	cmd := &cobra.Command{
> 		Use:     "diff -f FILENAME",
> 		Short:   i18n.T("Diff local configuration against live version"),
> ```

Is this correct since it is not the only thing that supported?

---

*[pkg/kubectl/cmd/diff.go, line 64 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KwaGcZ-4-Qe6aHWJo3l:-KwaGcZ-4-Qe6aHWJo3m:b-ldqvb) ([raw file](https://github.com/kubernetes/kubernetes/blob/42a1831ea571eca3af6ea889bbf02f9553a3160f/pkg/kubectl/cmd/diff.go#L64)):*
> than contains the configuration to diff

How about just `contains the configuration to diff`

---

*[pkg/kubectl/cmd/diff.go, line 83 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KwaGszQ08RssVtllCue:-KwaGszQ08RssVtllCuf:b-i6edhl) ([raw file](https://github.com/kubernetes/kubernetes/blob/42a1831ea571eca3af6ea889bbf02f9553a3160f/pkg/kubectl/cmd/diff.go#L83)):*
> ```Go
> 
> 	cmd := exec.Command(diff[0], diff[1:]...)
> 	out, err := cmd.StdoutPipe()
> ```

Why not just set the STDOUT and STDERR on the command to the writer?  Also, probably want stdout and stderr right?

---

*[pkg/kubectl/cmd/diff.go, line 101 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KwaUWjTmpINBBRMw3gf:-KwaUWjTmpINBBRMw3gg:b-xb6pap) ([raw file](https://github.com/kubernetes/kubernetes/blob/42a1831ea571eca3af6ea889bbf02f9553a3160f/pkg/kubectl/cmd/diff.go#L101)):*
> ```Go
> }
> 
> func RunDiff(f cmdutil.Factory, cmd *cobra.Command, stdout io.Writer, options *DiffOptions, args []string) error {
> ```

can we decouple the command related stuff from the business logic so that this can be executed without direct deps on the args / flags / env / stdin / stdout?

---

*[pkg/kubectl/cmd/diff.go, line 133 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KwaUH5KnKDu51yjpaNl:-KwaUH5KnKDu51yjpaNm:bau0rgp) ([raw file](https://github.com/kubernetes/kubernetes/blob/42a1831ea571eca3af6ea889bbf02f9553a3160f/pkg/kubectl/cmd/diff.go#L133)):*
> ```Go
> 	}
> 
> 	r := f.NewBuilder().
> ```

there is a lot of business logic in setting this up.  wdyt of pulling it into its own function

---

*[pkg/kubectl/cmd/diff.go, line 191 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KwaTro7mb04nM62UhYQ:-KwaTro7mb04nM62UhYR:bpxa02e) ([raw file](https://github.com/kubernetes/kubernetes/blob/42a1831ea571eca3af6ea889bbf02f9553a3160f/pkg/kubectl/cmd/diff.go#L191)):*
> ```Go
> 	}
> 
> 	return from, to, nil
> ```

validate the values

---

*[pkg/kubectl/cmd/diff.go, line 279 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KwaVHUlDyeoGi7FhxPQ:-KwaVHUlDyeoGi7FhxPR:be8dxc8) ([raw file](https://github.com/kubernetes/kubernetes/blob/42a1831ea571eca3af6ea889bbf02f9553a3160f/pkg/kubectl/cmd/diff.go#L279)):*
> ```Go
> }
> 
> type Object struct {
> ```

I think we are going to want to reuse this for the apply code right?

---

*[pkg/kubectl/cmd/diff.go, line 283 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KwaVoeS1OZ70d8GgINd:-KwaVoeS1OZ70d8GgINe:bj3ghpu) ([raw file](https://github.com/kubernetes/kubernetes/blob/42a1831ea571eca3af6ea889bbf02f9553a3160f/pkg/kubectl/cmd/diff.go#L283)):*
> ```Go
> }
> 
> func (obj Object) Local(encoder runtime.Encoder) (map[string]interface{}, error) {
> ```

I think the encoder should go in the `Object` struct

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920)*

**Comment by apelisse** [comment]
Just pushed a new version with lots of improvements. I still need to fix go vet though.

**Comment by pwittrock** [comment]
@apelisse Sweet!

**Comment by pwittrock** [comment]
*[pkg/kubectl/cmd/diff.go, line 62 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KxA3H1L6gL29myteWhw:-KxA3H1Ms6iWLFrRVe6h:bnlm2z5) ([raw file](https://github.com/kubernetes/kubernetes/blob/011a45cde726748a9790ef501d22658bb7909641/pkg/kubectl/cmd/diff.go#L62)):*
> ```Go
> 		cat service.yaml | kubectl alpha diff -f - MERGED
> 
> 		# Or specify both versions
> ```

FWIW. This is also specifying 2 files

Can you add an example for doing a directory?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920)*

**Comment by pwittrock** [comment]
@apelisse Looks like there are still some issues with the checks

**Comment by apelisse** [comment]
test ran before #54441 got merged. Let's try again

/retest

**Comment by pwittrock** [comment]
Review status: 0 of 8 files reviewed at latest revision, 14 unresolved discussions.

---

*[pkg/kubectl/cmd/diff.go, line 229 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KxBZjJaolo714AQpR0E:-KxBZjJaolo714AQpR0F:b-j87y6c) ([raw file](https://github.com/kubernetes/kubernetes/blob/011a45cde726748a9790ef501d22658bb7909641/pkg/kubectl/cmd/diff.go#L229)):*
> ```Go
> }
> 
> // CreateDirectory does create the actual disk directory, and return a
> ```

I don't really understand this comment.  This just returns a new Directory wrapping a temporary directory on disk.

---

*[pkg/kubectl/cmd/diff.go, line 252 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KxBZazHFax_ixzkaCzO:-KxBZazHFax_ixzkaCzP:b-bjucmr) ([raw file](https://github.com/kubernetes/kubernetes/blob/011a45cde726748a9790ef501d22658bb7909641/pkg/kubectl/cmd/diff.go#L252)):*
> ```Go
> }
> 
> // Object is an interface that let's you retrieve multiple version of
> ```

This comment could be more descriptive.

---

*[pkg/kubectl/cmd/diff.go, line 265 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KxBZSsxCzh8TJ9yVMOS:-KxBZSsxCzh8TJ9yVMOT:b-n1w7z2) ([raw file](https://github.com/kubernetes/kubernetes/blob/011a45cde726748a9790ef501d22658bb7909641/pkg/kubectl/cmd/diff.go#L265)):*
> ```Go
> // InfoObject is an implementation of the Object interface. It gets all
> // the information from the Info object.
> type InfoObject struct {
> ```

This is cool

---

*[pkg/kubectl/cmd/diff.go, line 371 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KxBZJ6mx4OZ6v2SR3N2:-KxBZJ6mx4OZ6v2SR3N3:bddei1n) ([raw file](https://github.com/kubernetes/kubernetes/blob/011a45cde726748a9790ef501d22658bb7909641/pkg/kubectl/cmd/diff.go#L371)):*
> ```Go
> }
> 
> // Diff diffs to versions of a specific object, and print both versions to directories.
> ```

`diffs two versions`?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920)*

**Comment by pwittrock** [comment]
Review status: 0 of 8 files reviewed at latest revision, 16 unresolved discussions.

---

*[pkg/kubectl/cmd/diff.go, line 179 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KxBa-kvCq-IerI5GzQq:-KxBa-kvCq-IerI5GzQr:b-bn0zff) ([raw file](https://github.com/kubernetes/kubernetes/blob/011a45cde726748a9790ef501d22658bb7909641/pkg/kubectl/cmd/diff.go#L179)):*
> ```Go
> 
> // DiffVersion gets the proper version of objects, and aggregate them into a directory.
> type DiffVersion struct {
> ```

`DiffSource` might be a better name.  The notion of Version we use for APIs is very different than how we use it here.

---

*[pkg/kubectl/cmd/diff.go, line 211 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KxB_KdjyzFXm9KcUmrQ:-KxB_KdjyzFXm9KcUmrR:bqshxfg) ([raw file](https://github.com/kubernetes/kubernetes/blob/011a45cde726748a9790ef501d22658bb7909641/pkg/kubectl/cmd/diff.go#L211)):*
> ```Go
> 
> // Print prints the object using the printer into a new file in the directory.
> func (v *DiffVersion) Print(obj Object, printer Printer) error {
> ```

I think printer should be part of DiffVersion.  It seems like a library dependency of DiffVersion, not an argument.  Then it could be allocated automatically and optionally overridden.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920)*

**Comment by pwittrock** [comment]
Review status: 0 of 8 files reviewed at latest revision, 14 unresolved discussions.

---

*[pkg/kubectl/cmd/diff.go, line 141 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KxDjX6OzkkobnFqGZ_a:-KxDjX6OzkkobnFqGZ_b:byg7psn) ([raw file](https://github.com/kubernetes/kubernetes/blob/011a45cde726748a9790ef501d22658bb7909641/pkg/kubectl/cmd/diff.go#L141)):*
> ```Go
> func (d *DiffProgram) getCommand(args ...string) exec.Cmd {
> 	diff := ""
> 	if envDiff := os.Getenv("KUBERNETES_EXTERNAL_DIFF"); envDiff != "" {
> ```

Instead of reading this from the environment here, include it in the struct having already been read in.  This helps decouple unrelated implementation details.  (e.g. where the diff command override comes from, and how the diff is executed)

---

*[pkg/kubectl/cmd/diff.go, line 254 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KxDeQwVMa_7MhyffqNp:-KxDeQwVMa_7MhyffqNq:b-i4kvkh) ([raw file](https://github.com/kubernetes/kubernetes/blob/011a45cde726748a9790ef501d22658bb7909641/pkg/kubectl/cmd/diff.go#L254)):*
> ```Go
> // Object is an interface that let's you retrieve multiple version of
> // it.
> type Object interface {
> ```

Lets put this stuff in a separate package so it can't depend upon the cobra command.  Maybe `pkg/kubectl/cmd/apply/resource`  Setup the Bazel rule so it can't depend on cobra.  Add a `doc.go` file to the new package with usage examples.  I would like each package in kubectl to have a `doc.go` with examples, and writing these will be a good new contributor task, but we need some canonical doc.go examples.

---

*[pkg/kubectl/cmd/diff.go, line 255 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KxDf7Sq2tJjqgGVqTeq:-KxDf7Sq2tJjqgGVqTer:brrywoa) ([raw file](https://github.com/kubernetes/kubernetes/blob/011a45cde726748a9790ef501d22658bb7909641/pkg/kubectl/cmd/diff.go#L255)):*
> ```Go
> // it.
> type Object interface {
> 	Local() (map[string]interface{}, error)
> ```

Document these well.  Assume they will used by someone who didn't write them.

---

*[pkg/kubectl/cmd/diff.go, line 350 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KxDhTFkDZm1UlvLERaR:-KxDhTFkDZm1UlvLERaS:b7rbeuv) ([raw file](https://github.com/kubernetes/kubernetes/blob/011a45cde726748a9790ef501d22658bb7909641/pkg/kubectl/cmd/diff.go#L350)):*
> ```Go
> 
> // Differ creates two DiffVersion and diffs them.
> type Differ struct {
> ```

Lets put this in `pkg/kubectl/cmd/apply/diff` and make it private to the package.  Only expose the version that runs on multiple resources.

---

*[pkg/kubectl/cmd/diff.go, line 396 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KxDhf0rjCATyjdIoVE6:-KxDhf0rjCATyjdIoVE7:bcrq70p) ([raw file](https://github.com/kubernetes/kubernetes/blob/011a45cde726748a9790ef501d22658bb7909641/pkg/kubectl/cmd/diff.go#L396)):*
> ```Go
> // diff, and find each Info object for each files, and runs against the
> // differ.
> func RunDiff(f cmdutil.Factory, diff *DiffProgram, options *DiffOptions, from, to string) error {
> ```

`cmdutil.Factory` is a bit of a mess to use.  I obfuscates the true dependencies of a function.  How about making a struct type that takes the dependencies from cmdutil.Factory as fields, and make this a function of that.  Maybe call it `DiffResources` or `MultiDiffer`.  Move this to `pkg/kubectl/cmd/apply/diff`

---

*[pkg/kubectl/cmd/diff_test.go, line 61 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KxDiqNF0GiokLoupawv:-KxDiqNF0GiokLoupaww:bpsc8s1) ([raw file](https://github.com/kubernetes/kubernetes/blob/14ec200e4e7172db0bd35f155c4befaf6d9261b0/pkg/kubectl/cmd/diff_test.go#L61)):*
> ```Go
> }
> 
> func TestArguments(t *testing.T) {
> ```

Later, we should standardize on either ginkgo or this.  I tend to prefer ginkgo since it provides good matching / expect libraries, and good support for test harnesses.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920)*

**Comment by pwittrock** [comment]
Review status: 0 of 8 files reviewed at latest revision, 15 unresolved discussions.

---

*[pkg/kubectl/cmd/diff.go, line 283 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KxDpbQlMcA7ZwOSdhwy:-KxDpbQlMcA7ZwOSdhwz:btupe2i) ([raw file](https://github.com/kubernetes/kubernetes/blob/011a45cde726748a9790ef501d22658bb7909641/pkg/kubectl/cmd/diff.go#L283)):*
> ```Go
> 
> func (obj InfoObject) Local() (map[string]interface{}, error) {
> 	data, err := runtime.Encode(obj.Encoder, obj.Info.VersionedObject)
> ```

Can we encapsulate the encoding / decoding logic in a simple library / interface.  This is black magic to most of us.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920)*

**Comment by pwittrock** [comment]
Review status: 0 of 8 files reviewed at latest revision, 15 unresolved discussions.

---

*[pkg/kubectl/cmd/diff.go, line 283 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KxDpbQlMcA7ZwOSdhwy:-KxDpkPAwbMnUUxdC0nC:bszt805) ([raw file](https://github.com/kubernetes/kubernetes/blob/011a45cde726748a9790ef501d22658bb7909641/pkg/kubectl/cmd/diff.go#L283)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary><blockquote>

Can we encapsulate the encoding / decoding logic in a simple library / interface.  This is black magic to most of us.
</blockquote></details>

Maybe `pkg/kubectl/object`?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920)*

**Comment by pwittrock** [comment]
Review status: 0 of 8 files reviewed at latest revision, 16 unresolved discussions.

---

*[pkg/kubectl/cmd/diff.go, line 225 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KxDqBAZuIlS7WUJZVeN:-KxDqBAZuIlS7WUJZVeO:b-y0ml62) ([raw file](https://github.com/kubernetes/kubernetes/blob/011a45cde726748a9790ef501d22658bb7909641/pkg/kubectl/cmd/diff.go#L225)):*
> ```Go
> 
> // Directory creates a new temp directory, and allows to easily create new files.
> type Directory struct {
> ```

This seems generally useful.  Put it in a shared location?  `pkg/kubectl/directory`?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920)*

**Comment by pwittrock** [comment]
Review status: 0 of 8 files reviewed at latest revision, 17 unresolved discussions.

---

*[pkg/kubectl/cmd/diff.go, line 162 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KxDqRjx6wVXocosu9Si:-KxDqRjx6wVXocosu9Sj:bhekzd0) ([raw file](https://github.com/kubernetes/kubernetes/blob/011a45cde726748a9790ef501d22658bb7909641/pkg/kubectl/cmd/diff.go#L162)):*
> ```Go
> 
> // Printer is used to print an object.
> type Printer struct{}
> ```

This is generally useful, and should go in its own library with encode / decode helper functions.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920)*

**Comment by pwittrock** [comment]
Review status: 0 of 8 files reviewed at latest revision, 18 unresolved discussions.

---

*[pkg/kubectl/cmd/diff.go, line 325 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-KxDqeOGcbk1p6wBoRpA:-KxDqeOGcbk1p6wBoRpB:b3ygu9g) ([raw file](https://github.com/kubernetes/kubernetes/blob/011a45cde726748a9790ef501d22658bb7909641/pkg/kubectl/cmd/diff.go#L325)):*
> ```Go
> 		return nil, err
> 	}
> 	result, err := elmt.Merge(strategy.Create(strategy.Options{}))
> ```

probably want to pass in strategy.Options to the function so callers can tweak behavior.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920)*

**Comment by pwittrock** [comment]
Lets try to figure out how to really keep code modular and encapsulated using packages and multiple files.  As a rule of thumb, most non-test files should be less than 200 lines.  If they are longer, we should look closely to see if everything in the file needs to be tightly coupled.

---

Review status: 0 of 8 files reviewed at latest revision, 18 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-:-KxDrJdFcGDNwz32B2xQ:b-rbj0r4)*

**Comment by pwittrock** [comment]
Lets start thinking about how our code would look with something like.

https://godoc.org/github.com/facebookgo/inject#example-package

---







*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53920#-:-KxDsRyVv4rblcoHwJhf:b-hvzk9b)*

**Comment by pwittrock** [comment]
Let fix the nits around structure in a follow up

File: pkg/kubectl/cmd/diff.go [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+func (obj InfoObject) toMap(data []byte) (map[string]interface{}, error) {
+	m := map[string]interface{}{}
+	if len(data) == 0 {
+		return m, nil
+	}
+	err := json.Unmarshal(data, &m)
+	return m, err
+}
+
+func (obj InfoObject) Local() (map[string]interface{}, error) {
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/diff.go]: @apelisse this assumption that VersionedObject is "local" seems questionable.  Can you explain why you're treating it this way?  Are you setting VersionedObject to represent something different than what most builders are assuming?

Comment (apelisse) [reply pkg/kubectl/cmd/diff.go]: Someone told me that might do what I want, but I have no idea what I'm doing. Feedback is welcome :-)

Comment (redbaron) [reply pkg/kubectl/cmd/diff.go]: @smarterclayton , @apatil , I dont know if was broken from the beginning or this proposed change broke it, but in 1.9.2 kubectl diff command shows nothing. Quick check finds that LOCAL and LIVE files written by differ are byte-to-byte the same, which is not surprising that currently `obj.Local()` and `obj.Live()` serialize very same `obj.Info.Object`.

Before https://github.com/kubernetes/kubernetes/pull/55660 (== in this original PR) `Local()` was serializing `obj.Info.VersionedObject` , which doesn't look very local to me either, but I have no idea about these internals.

Here is how I call diff during my tests:
```
kubectl -n kube-system apply -f  manifests/kube-dns-deployment.yaml
vim manifests/kube-dns-deployment.yaml (add label)
kubectl -n kube-system alpha diff -f manifests/kube-dns-deployment.yaml
```

**Pull request merged**